docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+
+Platform or ML engineers can tweak existing blueprints, based on their requirements. For example,
+using custom configurations.
+
+The Kubeflow Training Operator can maintain more Training Runtimes when the community is ready to
+support them. For example, runtimes for [Jax](https://jax.readthedocs.io/en/latest/index.html) or
+[MLX](https://ml-explore.github.io/mlx/build/html/index.html). We will support PyTorch and MPI runtimes.
+After initial implementation, we will support TensorFlow, XGboost, and PaddlePaddle runtimes, but
+it is out of scope for this KEP.
+
+The `TrainingRuntime` is immutable, and so to make a change, a new version of the `TrainingRuntime`
+must be created and then the user must change the `TrainJob` to point to the new version.
+This provides control as to how changes to runtimes propagate to existing training jobs.
+For example, when training is running for a long time (e.g. 1-2 months).
+
+In the future implementation, we will introduce a revision control mechanism similar to
+[Kubernetes Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#updating-a-deployment)
+to control versions of `TrainingRuntime` and enable rolling updates.
+
+We are going to create two CRDs: `TrainingRuntime` and `ClusterTrainingRuntime`. These runtimes have
+exactly the same APIs, but the first one is the namespace-scoped, the second is the cluster-scoped.
+If `trainingRuntimeRef` in `TrainJob` has the namespace, controller will use the `TrainingRuntime`,
+otherwise it will use the `ClusterTrainingRuntime`.
+
+```golang
+type TrainingRuntime struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Framework specific parameters.
+    MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+    // Number of nodes to execute training.
+    NumNodes int `json:"numNodes,omitempty"`
+
+    // JobSet spec.
+    JobSetSpec *batchv1.JobSetSpec `json:",inline"`
+
+    // For gang-scheduling using volcano or scheduler plugins, supported for all frameworks.
+    GangScheduler *GangScheduler `json:"gangScheduler,omitempty"`
+}
+
+// One of the specs can be selected.
+type MLSpec struct {
+
+    // Custom Spec for Torch
+    TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+    // Custom Spec for MPI
+    MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+```
+
+### The Gang Scheduler API
+
+Gang scheduler plugin is used to create the appropriate `PodGroup` for Volcano or scheduler plugins.
+
+```golang
+type GangScheduler struct {
+    // Plugin for gang scheduling.
+    Plugin *GangSchedulerPlugin `json:plugin,omitempty"`
+
+    // Time threshold to schedule PodGroup for gang scheduling.
+    ScheduleTimeoutSeconds string `json:scheduleTimeoutSeconds,omitempty"`
+}
+
+type GangSchedulerPlugin string
+
+const (
+    GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+    GangSchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"
```

Comment:
```suggestion
    SchedulerPlugins       GangSchedulerPlugin = "scheduler-plugins"
```

Actually, this is the coscheduling scheduler-plugins.

Reply:
If we initially support only Coscheduling: https://github.com/kubernetes-sigs/scheduler-plugins?tab=readme-ov-file#plugins, should we call it: `GangSchedulerPluginCoscheduling` ?

Reply:
Indeed, we support only the coscheduling plugin. I'm fine with either name.

Note that the scheduler-plugins support the gang scheduling only by the coscheduling plugin.

Reply:
Makes sense, I think it is better to be more declarative here.